### ISSUE TITLE
Secondary structure assignment could be done taking all symmetry partners into account

### ISSUE THREAD

**Issue by josemduarte** [biojava/biojava#454]
It is very usual that beta sheets form through symmetry partners in the crystal (be them biological contacts or not). It would be nice if our dssp-like secondary assignment implementation would do assignment taking into account all contacts from symmetry partners. We do have the code around to calculate the contacts (`org.biojava.nbio.structure.xtal.CrystalBuilder`) so this should be doable with our current code.

An extreme example of a case where this is important is 4ufq, where 1 chain in AU + 2 symmetry partner interleave together to form 3 beta sheets, with subsequent strands coming from alternate symmetry partners.

**Comment by abradle** [comment]
I'm not so sure about this. I wonder if it might add more noise than signal. And for very large structures it will increase the compute time quite significantly.

**Comment by lafita** [comment]
In my opinion it is only important to take into account the contacts in the **biological assembly**.

Secondary Structure (SS) formed at the crystal contacts is artificial, and probably without the contacts these regions would be disordered. Is it biologically more meaningful to report that they are disordered or that they can form SS?
I agree with @abradle that it adds more noise, because the biological SS and the crystal induced SS in these regions would not be distinguishable. 

However, @josemduarte's point is important, since **dssp** (and our implementation of it) calculates the SS induced between chains in the asymmetric unit (AU). This is probably the worst solution, since it takes into account some crystal contacts, while leaving some others out (not consistent).

A better approach would be, when the **biological assembly** is known, take into account only the contacts between assembly partners. These solution adds more freedom to the user, who can specify which chains (and symmetry partners) will be included in the SS assignment. For that we need a better way to represent symmetry partners in **biological assemblies** (see #220).

**Comment by josemduarte** [comment]
I agree to some extent that it'd be good to do it on bioassemblies only. But there are 2 problems with that:
1. bioassemblies are often badly annotated (so we could miss SS assignments that are important)
2. many times an entry is annotated with multiple bioassemblies

Because of number 2 we would need to run the SS detection on ALL bioassemblies provided. That introduces the risk of assigning different SS to different chains of the same entity (whenever a certain assembly is annotated for a chain but not for another of the same entity, which does happen from time to time). 

In order to avoid all those problems, I would advocate to do the SS assignment for all contacts in the crystal, so that there's no interpretation/bias introduced by the bioassemblies (which are interpretations of the crystal). 

Regarding your point that a SS in a crystal contact is "artificial", I'd defend that they are only as artificial as the crystal structure is. We can't say that they are artifacts or not without extra information from other experiments. But that applies to many things in the crystal structure and not only to SS assignments.

**Comment by larsonmattr** [comment]
It would be helpful for the SS assignment to recognize the sheets formed
between symmetry related molecules. Amyloid fibril structures are an
example of a non-artificial pairing of strands between the symmetry-related
molecules.

This might require a way to describe interactions with particular symmetry
related molecules - I believe something like this was also needed for
_struct_conn entries.

On Wed, Apr 20, 2016 at 2:07 PM, Jose Manuel Duarte <
notifications@github.com> wrote:

> I agree to some extent that it'd be good to do it on bioassemblies only.
> But there are 2 problems with that:
> 1. bioassemblies are often badly annotated (so we could miss SS
> assignments that are important)
> 2. many times an entry is annotated with multiple bioassemblies
>
> Because of number 2 we would need to run the SS detection on ALL
> bioassemblies provided. That introduces the risk of assigning different SS
> to different chains of the same entity (whenever a certain assembly is
> annotated for a chain but not for another of the same entity, which does
> happen from time to time).
>
> In order to avoid all those problems, I would advocate to do the SS
> assignment for all contacts in the crystal, so that there's no
> interpretation/bias introduced by the bioassemblies (which are
> interpretations of the crystal).
>
> Regarding your point that a SS in a crystal contact is "artificial", I'd
> defend that they are only as artificial as the crystal structure is. We
> can't say that they are artifacts or not without extra information from
> other experiments. But that applies to many things in the crystal structure
> and not only to SS assignments.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/biojava/biojava/issues/454#issuecomment-212561166>
>



-- 
Matt Larson, PhD
Madison, WI  53705 U.S.A.

**Comment by josemduarte** [comment]
I think we should have this as a parameter: users can request whether to use symmetry mates or not for SS assignments. Then the user can decide.

**Comment by andreasprlic** [comment]
Just to add, there are many things that could be improved on top of the original DSSP. E.g. usually helices are too short!

**Comment by lafita** [comment]
Some of the things that could be improved in SS assignment were discussed in #112. We decided to stick with the **DSSP** conventions, since it has been the standard. 

However, I think it is a good idea to provide a more general (customizable) version of the algorithm, since this was one of the purposes of having SS assignment in **biojava**. My suggestion is to create a **parameter bean** `Class` that is passed either to the `SecStrucCalc` constructor or to the method `SecStrucCalc.calculate()`. Some of the parameters that can be included are:
- include symmetry partners
- helix length
- H bond energy threshold
- H placement method

We could either make the **DSSP** options as the default or create a static method that returns a `Parameters` object with the **DSSP** options.
